pkg/monitor/launch/launcher.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package launch
+
+import (
+	"bufio"
+	"encoding/json"
+	"os"
+	"os/exec"
+	"sync"
+
+	"github.com/cilium/cilium/api/v1/models"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const targetName = "cilium-node-monitor"
+
+var stateMutex sync.Mutex
+
+// NodeMonitor is used to wrap the node executable binary.
+type NodeMonitor struct {
+	Arg     string
+	process *os.Process
+	state   *models.MonitorStatus
+}
+
+// Run starts the node monitor.
+func (nm *NodeMonitor) Run() {
+	for {
+		stateMutex.Lock()
+		cmd := exec.Command(targetName, nm.Arg)
+		stateMutex.Unlock()
+		stdout, _ := cmd.StdoutPipe()
+
+		if err := cmd.Start(); err != nil {
+			log.Errorf("cmd.Start(): %s", err)
+		}
+		nm.process = cmd.Process
+		go func() { cmd.Wait() }()
+
+		r := bufio.NewReader(stdout)
+		for nm.process != nil {
+			l, _ := r.ReadString('\n')
+			var tmp *models.MonitorStatus
+			if err := json.Unmarshal([]byte(l), &tmp); err != nil {
+				continue
+			}
+
+			stateMutex.Lock()
+			nm.state = tmp
+			stateMutex.Unlock()
+		}
+	}
+}
+
+// Restart stops the node monitor and then runs it again.
+func (nm *NodeMonitor) Restart(arg string) {
+	stateMutex.Lock()
+	defer stateMutex.Unlock()
+	nm.Arg = arg
+
+	if nm.process == nil {
+		return
+	}
+	if err := nm.process.Kill(); err != nil {
+		log.Errorf("process.Kill(): %s", err)
+	}
+	nm.process = nil
+}
+
+// StateLocked returns the monitor status.
+func (nm *NodeMonitor) StateLocked() *models.MonitorStatus {
```

Comment:
We typically use the `Locked()` suffix to indicate that the structure must already be locked.